### PR TITLE
feat: [WD-13153] Added cluster form and Join Token creation functiona…

### PR DESCRIPTION
**Pull request by Kxiru**
…lity

**Work Done:**
- Added a new Create Cluster / Token page.
- Added url / CreateCluster component to App.tsx
- Added an API call to create a Token.
- A Formik form to accept Token attributes.
- Minimal styling.
- Input validation
- Added a Version status bar at the bottom of the page.

image

image

Minimal Input Validation:
image
image

### CODE REVIEW

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  const getPayload = (values: CreateClusterFormValues) => {
+    return {
+      ...newTokenPayload(values),
+    };
+  };
+
+  const formik = useFormik<CreateClusterFormValues>({
+    initialValues: {
+      site_name: "testsitename",
+      expiry: "2029-07-16T18:06:48Z",
```
Comment (edlerd) [inline ui/src/pages/clusters/CreateCluster.tsx]: The default format is not valid. I think this should 24 hours from now.

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  const getPayload = (values: CreateClusterFormValues) => {
+    return {
+      ...newTokenPayload(values),
+    };
+  };
+
+  const formik = useFormik<CreateClusterFormValues>({
+    initialValues: {
+      site_name: "testsitename",
```
Comment (edlerd) [inline ui/src/pages/clusters/CreateCluster.tsx]:
```suggestion
      site_name: "",
```

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  const submit = (values: CreateClusterFormValues) => {
+    const newTokenPayload = getPayload(values);
+    navigate(`/ui/sites/tokens`);
+
+    createToken(JSON.stringify(newTokenPayload)).catch((e: Error) => {
+      if (e.message === "Unable to create new Token") {
+        return;
+      }
+    });
```
Comment (edlerd) [inline ui/src/pages/clusters/CreateCluster.tsx]:
```suggestion
    createToken(JSON.stringify(newTokenPayload))
    .then(() => {
      navigate(`/ui/sites/tokens`);
    })
    .catch((e: Error) => {
      if (e.message === "Unable to create new Token") {
        return;
      }
    });
```
We should only navigate once the creation is finished.
Also should add a `.finally` that is invalidating the token cache so we can see the newly created token immediately

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  };
+
+  const getPayload = (values: CreateClusterFormValues) => {
+    return {
+      ...newTokenPayload(values),
+    };
+  };
```
Comment (edlerd) [inline ui/src/pages/clusters/CreateCluster.tsx]: I think this can be done inline in `submit`, no need for an extra function.

**Comment by mas-who** [comment]
Some minor comments from QA:
 1. The buttons at the bottom of the form should align with the edge of the form 
Screenshot from 2024-07-17 17-57-37
**See LXD-UI for ref**
Screenshot from 2024-07-17 18-06-59

2. I think we should do validation for the expiry date that it can't be set to in the past.

File: ui/src/App.tsx [hunk 5]
```diff
@@ -33,6 +34,7 @@ const App: FC = () => {
         <Route path="/ui/login" element={<Login />} />
         <Route path="/ui/sites" element={<ClusterList />} />
         <Route path="/ui/sites/:activeTab" element={<ClusterList />} />
+        <Route path="/ui/sites/create" element={<CreateCluster />} />
```
Comment (mas-who) [inline ui/src/App.tsx:37]: please add this path to `internal/api/ui.go` file in the `uiServeRoutes` array

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+} from "./ClusterCreateDetailsForm";
+import FormFooterLayout from "components/forms/FormFooterLayout";
+import { createToken } from "api/tokens";
+import * as Yup from "yup";
+
+const CreateCluster: FC = () => {
+  const navigate = useNavigate();
+
+  const ClusterSchema = Yup.object().shape({
+    expiry: Yup.date().optional(),
```
Comment (mas-who) [inline ui/src/pages/clusters/CreateCluster.tsx]: validate expiry not in the past

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  };
+
+  const formik = useFormik<CreateClusterFormValues>({
+    initialValues: {
+      site_name: "testsitename",
+      expiry: "2029-07-16T18:06:48Z",
+    },
+    validationSchema: ClusterSchema,
+    onSubmit: (values) => {
+      console.log("test submission worked");
```
Comment (mas-who) [inline ui/src/pages/clusters/CreateCluster.tsx]: remember to remove after testing

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      submit(values);
+    },
+  });
+
+  return (
+    <BaseLayout
+      title="Create a new join Token"
+      contentClassName="create-cluster"
+    >
+      <Row>
```
Comment (mas-who) [inline ui/src/pages/clusters/CreateCluster.tsx]: I don't think this `Row` is necessary here.

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        <Form onSubmit={formik.handleSubmit} className="form">
+          <Row className="form-contents">
+            <Col size={12}>
+              {/* <NotificationRow /> */}
```
Comment (mas-who) [inline ui/src/pages/clusters/CreateCluster.tsx]: remove comment for now. This will be related to notifications in a later PR

File: ui/src/sass/_forms.scss [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+      .content-details {
+        max-width: 54rem;
+      }
+    }
+  }
+
+  .bottom-btns {
+    margin-top: 1rem;
+    max-width: 72rem;
```
Comment (mas-who) [inline ui/src/sass/_forms.scss]: this may need to be tweaked to get buttons align with the form

File: ui/src/components/ScrollableContainer.tsx [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  children: ReactNode;
+  dependencies: DependencyList;
+  belowIds?: string[];
+  className?: string;
+}
+
+const ScrollableContainer: FC<Props> = ({
+  dependencies,
+  children,
+  belowIds = ["status-bar"],
```
Comment (mas-who) [inline ui/src/components/ScrollableContainer.tsx:19]: "status-bar" is a very specific id in lxd-ui. Let's remove this and default `belowIds` to just an empty list

File: ui/src/components/ScrollableForm.tsx [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+interface Props {
+  children: ReactNode;
+}
+
+const ScrollableForm: FC<Props> = ({ children }) => {
+  const notify = useNotify();
+  return (
+    <ScrollableContainer
+      dependencies={[notify.notification]}
+      belowIds={["status-bar", "form-footer"]}
```
Comment (mas-who) [inline ui/src/components/ScrollableForm.tsx:14]: remove `status-bar`

File: ui/src/pages/clusters/CreateCluster.tsx (hunk 3 above)
Comment (mas-who) [reply ui/src/pages/clusters/CreateCluster.tsx]: I think in the success case, we also need to invalidate the `tokens` query key

**Review by mas-who — COMMENTED** [review]
Thanks for implementing this. Left some feedback :)

File: ui/src/components/ScrollableContainer.tsx (hunk 11 above)
Comment (edlerd) [reply ui/src/components/ScrollableContainer.tsx:19]: @mas-who is raising an interesting point. We might want to add a status bar that is displaying the version. Even though we don't have a version yet. We can hardcode "Version 0.1" for now.

Comment (Kxiru) [reply ui/src/components/ScrollableContainer.tsx:19]: Done!

**Comment by Kxiru** [comment]
Added a status bar at the bottom! So cute haha.
image

File: ui/src/components/StatusBar.tsx [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import { FC } from "react";
+import classnames from "classnames";
+
+interface Props {
+  className?: string;
+}
+
+const StatusBar: FC<Props> = ({ className }) => {
```
Comment (mas-who) [inline ui/src/components/StatusBar.tsx]: Let's keep things consistent and use [AppStatus](https://canonical.github.io/react-components/?path=/docs/components-applicationlayout-appstatus--docs) from react-components.

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+  const ClusterSchema = Yup.object().shape({
+    site_name: Yup.string().required("Site name is required"),
+    expiry: Yup.date()
+      .optional()
+      .min(new Date(), "Expiry date cannot be in the past"),
+  });
+
+  const submit = (values: CreateClusterFormValues) => {
+    const tokenPayload = newTokenPayload(values);
+    navigate(`/ui/sites/tokens`);
```
Comment (mas-who) [inline ui/src/pages/clusters/CreateCluster.tsx]: this navigate should be removed. We should only navigate away from this page in the success case

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+    createToken(JSON.stringify(tokenPayload))
```
Comment (mas-who) [inline ui/src/pages/clusters/CreateCluster.tsx:37]: for future reference, the success and failure case will need notifications in the follow up PR

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        }
+      })
+      .finally(() => {
+        void queryClient.invalidateQueries({
+          queryKey: [queryKeys.tokens],
+        });
+      });
+  };
+
+  const getExpiryDate = () => {
```
Comment (mas-who) [inline ui/src/pages/clusters/CreateCluster.tsx]: for functions that do not handle state, it is better to define them outside of components. Otherwise functions gets redefined unnecessarily per render cycle. We can follow the folder structure in lxd-ui here. Create a file in `/util` to hold helper functions related specifically for creating tokens e.g. `/util/createCluster.tsx`. Also I think this function can be renamed to `getDefaultExpiryDate`

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    return now.toISOString();
+  };
+
+  const formik = useFormik<CreateClusterFormValues>({
+    initialValues: {
+      site_name: "",
+      expiry: getExpiryDate(),
+    },
+    validationSchema: ClusterSchema,
+    onSubmit: (values) => {
```
Comment (mas-who) [inline ui/src/pages/clusters/CreateCluster.tsx]: when passing a value in the form of `(input) => func(input)`, you can just pass `func` instead

File: ui/src/pages/clusters/CreateCluster.tsx [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+      </Form>
+      <FormFooterLayout>
+        <Button appearance="base" onClick={() => navigate("/")}>
+          Cancel
+        </Button>
+        <ActionButton
+          appearance="positive"
+          loading={formik.isSubmitting}
+          disabled={!formik.isValid}
+          onClick={() => submit(formik.values)}
```
Comment (mas-who) [inline ui/src/pages/clusters/CreateCluster.tsx:82]: We should remove this `onClick`. and set the `type` attribute for this button to be `submit`. The form will handle the submission logic itself. Also this should prevent potentially sending multiple requests to the backend.

File: ui/src/sass/styles.scss [hunk 19]
```diff
@@ -47,3 +50,9 @@ body {
 .l-main .row {
   max-width: none;
 }
+
+.p-bottom-controls {
```
Comment (mas-who) [inline ui/src/sass/styles.scss]: what's the context for `p-bottom-controls`, why is this needed?

**Comment by mas-who** [comment]
Left some more comments, nice one implementing the status bar and getting the form buttons aligned :slightly_smiling_face:

File: ui/src/components/StatusBar.tsx (hunk 13 above)
Comment (Kxiru) [reply ui/src/components/StatusBar.tsx]: I used StatusBar from LXD-UI, is AppStatus what it should have been?

Comment (mas-who) [reply ui/src/components/StatusBar.tsx]: yeah status bar has the `l-status` class which belongs to the app layout family of things

File: ui/src/pages/clusters/CreateCluster.tsx (hunk 17 above)
Comment (Kxiru) [reply ui/src/pages/clusters/CreateCluster.tsx]: Do you mean it should be structured like this:

``` 
onSubmit: submit(values);
```

I think onSubmit needs to have an arrow function as before. Please could you clarify?

Comment (mas-who) [reply ui/src/pages/clusters/CreateCluster.tsx]: just `onSubmit: submit` should be fine

**Comment by Kxiru** [comment]
Reviews all addressed :+1:

File: ui/src/components/StatusBar.tsx [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+import { AppStatus } from "@canonical/react-components";
+
+interface Props {
+  className?: string;
+}
+
+const StatusBar: FC<Props> = () => {
+  return (
+    <AppStatus className="status-bar" id="status-bar">
+      <aside>
```
Comment (mas-who) [inline ui/src/components/StatusBar.tsx]: This `aside` isn't needed. `AppStatus` wraps children with an `aside` already

Comment (Kxiru) [reply ui/src/components/StatusBar.tsx]: Thanks for the heads up :)

File: ui/src/pages/clusters/ClusterCreateDetailsForm.tsx [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+          />
+          <Input
+            id="expiry"
+            name="expiry"
+            type="datetime-local"
+            label="Expiry Date"
+            placeholder="Enter Date"
+            onBlur={formik.handleBlur}
+            onChange={formik.handleChange}
+            value={formik.values.expiry}
```
Comment (edlerd) [inline ui/src/pages/clusters/ClusterCreateDetailsForm.tsx:49]: I think there is something off with this. The initial value is not displayed here. Might be a format mismatch. However, the initial value is sent in the payload on form submit and changing the date/time also is reflected in the payload. So it is solely about reflecting the initial value as an initial state of this input.

**Review by edlerd — DISMISSED** [review]
Found an issue with the initial state of the expiry.

One other idea now that we have token creation is to add an action column to the token table and add a "revoke" button that hits the token `DELETE` endpoint. We can also keep that as a follow up, though.

**Comment by Kxiru** [comment]
> Found an issue with the initial state of the expiry.
> 
> One other idea now that we have token creation is to add an action column to the token table and add a "revoke" button that hits the token `DELETE` endpoint. We can also keep that as a follow up, though.

Thanks for the comments. I also implemented the revoke button and API really quickly. Here it is!

image

File: ui/src/pages/clusters/ClusterTokenActions.tsx [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+
+const ClusterTokenActions: FC<Props> = ({ token }) => {
+  const navigate = useNavigate();
+  const queryClient = useQueryClient();
+
+  return (
+    <ActionButton
+      className={"u-no-margin"}
+      appearance="negative"
+      onClick={() => {
```
Comment (mas-who) [inline ui/src/pages/clusters/ClusterTokenActions.tsx]: it's usually a good idea to separate logic and presentation since TS mixed with component HTML is difficult to read. Would you please extract this callback into a `handleDeleteToken` function and pass it to the `onClick`?

File: ui/src/pages/clusters/ClusterTokenActions.tsx [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+        deleteToken(token.site_name)
+          .then(() => {
+            navigate(`/ui/sites/tokens`);
```
Comment (mas-who) [inline ui/src/pages/clusters/ClusterTokenActions.tsx]: I don't think we need to do a navigate after deleting the token. Also I'd remove the `finally` block and keep the query invalidation in the success case.

File: ui/src/pages/clusters/ClusterTokenActions.tsx (hunk 22 above)
Comment (Kxiru) [reply ui/src/pages/clusters/ClusterTokenActions.tsx]: Done :)

File: ui/src/pages/clusters/ApproveClusterButton.tsx [hunk 24]
```diff
@@ -28,6 +28,7 @@ const ApproveClusterButton: FC<Props> = ({ clusterName }) => {
 
   return (
     <ActionButton
+      className={"u-no-margin--bottom"}
```
Comment (edlerd) [inline ui/src/pages/clusters/ApproveClusterButton.tsx]:
```suggestion
      className="u-no-margin--bottom"
```

File: ui/src/pages/clusters/ClusterTokenActions.tsx [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+import { ActionButton } from "@canonical/react-components";
+import { deleteToken } from "api/tokens";
+import { Token } from "types/token";
+import { queryKeys } from "util/queryKeys";
+
+interface Props {
+  token: Token;
+}
+
+const ClusterTokenActions: FC<Props> = ({ token }) => {
```
Comment (edlerd) [inline ui/src/pages/clusters/ClusterTokenActions.tsx]:
```suggestion
const RevokeTokenButton: FC<Props> = ({ token }) => {
```
Maybe a better name for this component and file?

File: ui/src/pages/clusters/DeleteClusterButton.tsx [hunk 26]
```diff
@@ -28,6 +28,7 @@ const DeleteClusterButton: FC<Props> = ({ clusterName }) => {
 
   return (
     <ActionButton
+      className={"u-no-margin--bottom"}
```
Comment (edlerd) [inline ui/src/pages/clusters/DeleteClusterButton.tsx]:
```suggestion
      className="u-no-margin--bottom"
```

**Review by mas-who — DISMISSED** [review]

**Comment by mas-who** [comment]
LGTM :+1:

**Review by mas-who — APPROVED** [review]

**Review by edlerd — APPROVED** [review]
LGTM :+1:
